src/main/java/org/pvv/shufflegene/Traverse.java

```diff
@@ -0,0 +1,163 @@
+        Collections.copy(input, result);
+        return result;
+    }
+
+    /**
+     * Set the edge map and associated variables based on an input string.
+     *
+     * @param sequence nucleotide sequence.
+     */
+    private void setEdgeMap(String sequence) {
+        char[] nucleotides = sequence.toCharArray();
+        this.start = nucleotides[0];
+        this.end = nucleotides[nucleotides.length - 1];
+        this.length = nucleotides.length;
+
+        this.edgeMap = new HashMap<>();
+
+        for (int i = 0; i < nucleotides.length - 1; i++) {
+            ArrayList<Character> edges = this.edgeMap.getOrDefault(nucleotides[i], new ArrayList<>());
+            edges.add(nucleotides[i + 1]);
+            this.edgeMap.put(nucleotides[i], edges);
+        }
+
+    }
+
+    /**
+     * Remove the first occurrence of a given edge from the traverse.
+     *
+     * @param edge Pair of characters giving the start and end of the vertex.
+     * @return whether the removal was successful or not.
+     */
+    public boolean removeEdge(Pair<Character, Character> edge) {
+        return edgeMap.get(edge.getValue0()).remove(edge.getValue1());
+    }
+
+    /**
+     * Append an edge to the end of the traverse.
+     *
+     * @param edge Pair of characters giving the start and end of the vertex.
+     */
+    public void appendEdge(Pair<Character, Character> edge) {
+        edgeMap.get(edge.getValue0()).add(edge.getValue1());
+    }
+
+    /**
+     * Perform a deep copy of edgeMap, used by {@link #deepCopy()}
+     *
+     * @return a copy of edgeMap
+     */
+    private HashMap<Character, ArrayList<Character>> copyEdgeMap() {
+        return this.edgeMap
+                .entrySet()
+                .stream()
+                .collect(Collectors.toMap(
+                        Map.Entry::getKey,
+                        e -> copyList(e.getValue()),
+                        (v1, v2) -> {
+                            throw new IllegalStateException("Duplicate key");
+                        },
+                        HashMap::new
+                ));
+    }
+
+    /**
+     * Convert the traverse to a string. This method is unsafe, since it only works when the traverse
+     * is in a fully connected state.
+     *
+     * @return string with nucleotide sequence
+     */
+    public String unsafeToString() throws IllegalStateException {
+        StringBuilder result = new StringBuilder();
+        HashMap<Character, ArrayList<Character>> edgeMapCopy = copyEdgeMap();
+
+        char current = start;
+        result.append(start);
+
+        try {
+            for (int i = 1; i < this.length; i++) {
+                char next = edgeMapCopy.get(current).remove(0);
+                result.append(next);
+                current = next;
+            }
+        } catch (IndexOutOfBoundsException e) {
+            throw new IllegalStateException("Traverse is not fully connected and cannot be converted to a string.");
+        }
```

Comment:
Exceptions... Good use of making the exception specific.  Always catch early and handle because after the first jump the context for the issue is lost.  The problem then comes what to do about flagging the problem further up. If you need to throw up to the highest level because it's basically fatal and this process has to abort, unchecked exceptions are better than having checked ones cluttering the method signature and reading the control flow.
A better solution is an extension of avoiding primitives, and that's to create an error state for your custom types.  At the periphery of the module (inputs and outputs) you do have to deal with things like strings, so this is not a great example, but in another case you may have wanted something like an `InvalidNucleotideSequence` which can then represent itself when asked to do things like turn into a string.